barretenberg/cpp/src/barretenberg/dsl/acir_format/bigint_constraint.cpp

```diff
@@ -1,33 +1,465 @@
+template void create_bigint_neg_constraint<GoblinUltraCircuitBuilder>(
+    const BigIntOperation& input, DSLBigInts<GoblinUltraCircuitBuilder>& dsl_bigint);
+template void create_bigint_mul_constraint<UltraCircuitBuilder>(const BigIntOperation& input,
+                                                                DSLBigInts<UltraCircuitBuilder>& dsl_bigint);
+template void create_bigint_mul_constraint<GoblinUltraCircuitBuilder>(
+    const BigIntOperation& input, DSLBigInts<GoblinUltraCircuitBuilder>& dsl_bigint);
+template void create_bigint_div_constraint<UltraCircuitBuilder>(const BigIntOperation& input,
+                                                                DSLBigInts<UltraCircuitBuilder>& dsl_bigint);
+template void create_bigint_div_constraint<GoblinUltraCircuitBuilder>(
+    const BigIntOperation& input, DSLBigInts<GoblinUltraCircuitBuilder>& dsl_bigint);
+
+template <typename Builder>
+void create_bigint_addition_constraint(const BigIntOperation& input, DSLBigInts<Builder>& dsl_bigint)
+{
+    switch (dsl_bigint.get_modulus_id(input.lhs)) {
+    case ModulusId::BN254_FR: {
+        auto lhs = dsl_bigint.bn254_fr(input.lhs);
+        auto rhs = dsl_bigint.bn254_fr(input.rhs);
+        dsl_bigint.set_bn254_fr(lhs + rhs, input.result);
+        break;
+    }
+    case ModulusId::BN254_FQ: {
+        auto lhs = dsl_bigint.bn254_fq(input.lhs);
+        auto rhs = dsl_bigint.bn254_fq(input.rhs);
+        dsl_bigint.set_bn254_fq(lhs + rhs, input.result);
+        break;
+    }
+    case ModulusId::SECP256K1_FQ: {
+        auto lhs = dsl_bigint.secp256k1_fq(input.lhs);
+        auto rhs = dsl_bigint.secp256k1_fq(input.rhs);
+        dsl_bigint.set_secp256k1_fq(lhs + rhs, input.result);
+        break;
+    }
+    case ModulusId::SECP256K1_FR: {
+        auto lhs = dsl_bigint.secp256k1_fr(input.lhs);
+        auto rhs = dsl_bigint.secp256k1_fr(input.rhs);
+        dsl_bigint.set_secp256k1_fr(lhs + rhs, input.result);
+        break;
+    }
+    case ModulusId::SECP256R1_FQ: {
+        auto lhs = dsl_bigint.secp256r1_fq(input.lhs);
+        auto rhs = dsl_bigint.secp256r1_fq(input.rhs);
+        dsl_bigint.set_secp256r1_fq(lhs + rhs, input.result);
+        break;
+    }
+    case ModulusId::SECP256R1_FR: {
+        auto lhs = dsl_bigint.secp256r1_fr(input.lhs);
+        auto rhs = dsl_bigint.secp256r1_fr(input.rhs);
+        dsl_bigint.set_secp256r1_fr(lhs + rhs, input.result);
+        break;
+    }
+    default: {
+        ASSERT(false);
+    }
+    }
+}
+
+template <typename Builder>
+void create_bigint_neg_constraint(const BigIntOperation& input, DSLBigInts<Builder>& dsl_bigint)
+{
+    switch (dsl_bigint.get_modulus_id(input.lhs)) {
+    case ModulusId::BN254_FR: {
+        auto lhs = dsl_bigint.bn254_fr(input.lhs);
+        auto rhs = dsl_bigint.bn254_fr(input.rhs);
+        dsl_bigint.set_bn254_fr(lhs - rhs, input.result);
```

Comment:
why is neg `lhs - rhs` ?

Reply:
I renamed it into sub.
Let me know if you prefer to have a neg, or both neg and sub .